source/distributed/tria.cc

```diff
@@ -567,6 +569,622 @@ namespace
+     * quadrant.
+     *
+     * @note We can handle a quadrant that is mapped by bi-linear or tri-linear
+     * mappings. Checking for a point in a cell of a curved domain required
+     * knowledge of the attached manifold.
+     *
+     * @return `int` interpreted as a C "bool". Zero means "stop the recursion".
+     * This can happen once we know the owner rank or if we know that a point
+     * does not belong to a quadrant.
+     */
+    static int
+    local_point_fn(typename internal::p4est::types<dim>::forest *  forest,
+                   typename internal::p4est::types<dim>::topidx    which_tree,
+                   typename internal::p4est::types<dim>::quadrant *quadrant,
+                   int                                             rank_begin,
+                   int                                             rank_end,
+                   void *                                          point);
+    /*******************************/
+    /*******************************/
+
+  private:
+    /**
+     * Simple struct to keep relavant data. Can be accessed though p4est's user
+     * pointer.
+     */
+    class QuadrantData
+    {
+    public:
+      QuadrantData();
+
+      void
+      set_cell_vertices(
+        typename internal::p4est::types<dim>::forest *  forest,
+        typename internal::p4est::types<dim>::topidx    which_tree,
+        typename internal::p4est::types<dim>::quadrant *quadrant,
+        const typename internal::p4est::types<dim>::quadrant_coord
+          quad_length_on_level);
+
+      void
+      initialize_mapping();
+
+      Point<dim>
+      map_real_to_unit_cell(const Point<dim> &p) const;
+
+      bool
+      is_in_this_quadrant(const Point<dim> &p) const;
+
+    private:
+      std::vector<Point<dim>> cell_vertices;
+
+      /*!
+       * Matrix holds coefficients mapping from this physical cell to unit
+       * cell.
+       */
+      FullMatrix<double> quadrant_mapping_matrix;
+
+      bool is_initialized_vertices;
+
+      bool is_fully_initialized;
+    };
+
+    /**
+     * Quadrant data to be filled upon call of `local_quadrant_fn`.
+     */
+    QuadrantData quadrant_data;
+
+    /**
+     * Owner ranks. Later passed to the return value.
+     */
+    std::vector<unsigned int> owner_rank;
+  }; // class PartitionSearch
+
+
+
+  template <int dim>
+  int
+  PartitionSearch<dim>::local_quadrant_fn(
+    typename internal::p4est::types<dim>::forest *  forest,
+    typename internal::p4est::types<dim>::topidx    which_tree,
+    typename internal::p4est::types<dim>::quadrant *quadrant,
+    int /* rank_begin */,
+    int /* rank_end */,
+    void * /* this is always nullptr */ point)
+  {
+    // point must be be nullptr here
+    AssertThrow(point == nullptr, dealii::ExcInternalError());
+
+    // we need the user pointer
+    PartitionSearch<dim> *this_object =
+      reinterpret_cast<PartitionSearch<dim> *>(forest->user_pointer);
```

Comment:
You should be able to use the `this` pointer instead.
```suggestion
    Assert(this == forest->user_pointer, ExcInternalError());
```